gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,351 @@
+    return true;
+  }
+
+  private void createDbIfNotExists(String dbName) {
+    Database db = new Database();
+    db.setName(dbName);
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(this.hiveDbRootDir.get() + Path.SEPARATOR + dbName + HIVE_DB_EXTENSION);
+
+    Lock lock = locks.get(dbName);
+    lock.lock();
+
+    try {
+      client.createDatabase(db);
+      log.info("Created database " + dbName);
+    } catch (AlreadyExistsException e) {
+      // Database already exists. Nothing to do.
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table getOrCreateTable(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      return this.client.getTable(dbName, tableName);
+    } catch (NoSuchObjectException e) {
+      return createTable(dbName, tableName, partition, sd);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createTable(String dbName, String tableName, Optional<HivePartition> partition, StorageDescriptor sd) {
+    Map<String, String> parameters = Maps.newHashMap();
+    parameters.put("EXTERNAL", Boolean.TRUE.toString());
+    Table table = new Table();
+    table.setParameters(parameters);
+    table.setDbName(dbName);
+    table.setTableName(tableName);
+
+    Preconditions.checkArgument(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType(TableType.EXTERNAL_TABLE.toString());
+
+    if (partition.isPresent()) {
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    table.setSd(sd);
+    try {
+      this.client.createTable(table);
+      log.info("Created table " + tableName + " in db " + dbName);
+    } catch (TException e) {
+      log.error(String.format("Unable to create table %s from db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    }
+    return table;
+  }
+
+  public boolean existsTable(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      return this.client.tableExists(dbName, tableName);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean existsPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
```

Comment:
Instead of locking on a single `HiveMetaStoreClient` object, I probably need to use a pool of `HiveMetaStoreClient` objects, since `HiveMetaStoreClient` doesn't seem to support concurrency.